ath79: add support for Archer A6 v2 EU/RU

L2jLiga opened the pull request:
This patch adds support for TP-Link Archer A6 v2 (EU/RU)

Hardware specification:
- SOC: Qualcomm QCA9563 @ 775MHz
- Flash: GigaDevice GD25Q64CSIG (8MiB)
- RAM: Zentel A3R1GE40JBF (128 MiB DDR2)
- Ethernet: Qualcomm QCA8337N: 4x 1Gbps LAN + 1x 1Gbps WAN
- Wireless:
  - 2.4GHz (bgn) QCA9563 integrated (3x3)
  - 5GHz (ac) Qualcomm QCA9886 (2x2)
- Button: 1x power, 1x reset, 1x wps
- LED: 6x LEDs: power, wlan2g, wlan5g, lan, wan, wps
- UART: There's no UART header on the board

Flash instructions:

Upload
openwrt-ath79-generic-tplink_archer-a6-v2-squashfs-factory.bin
via the router Web interface.

Flash instruction using tftp recovery:

1. Connect the computer to one of the LAN ports of the router
2. Set the computer IP to 192.168.0.66
3. Start a tftp server with the OpenWrt factory image in the
   tftp root directory renamed to ArcherA6v2_tp_recovery.bin.
4. Connect power cable to router, press and hold the reset
   button and turn the router on
5. Keep the reset button pressed until the WPS LED lights up
6. Wait ~150 seconds to complete flashing

According to the GPL source the non-EU variant has different
GPIOs assigned to some of the LEDs and buttons. The flash
layout might be different as well.

Signed-off-by: Andrey Chalkin <L2jLiga@gmail.com>

Replaces #3704 

According to https://github.com/openwrt/openwrt/pull/3704#issuecomment-748691641 created separated device for TP-Link Archer A6 v2 EU/RU

hauke commented on target/linux/ath79/dts/qca9563_tplink_archer-a6-v2.dts:
```diff
@@ -0,0 +1,121 @@
+#include "qca9563_tplink_archer-x6-v2.dtsi"
+
+/ {
+	compatible = "tplink,archer-a6-v2", "qca,qca9563";
```
If this is more or less the same file as target/linux/ath79/dts/qca9563_tplink_archer-c6-v2.dts you can also do this:

```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "qca9563_tplink_archer-c6-v2.dts"

/ {
	compatible = "tplink,archer-a6-v2", "qca,qca9563";
	model = "TP-Link Archer A6 v2 (EU/RU)";
};
```

If there are more changes needed you can add overwrite the settings.

adschm replied:
@hauke I don't really like deriving DTSes from other DTSes. It's always somewhat counter-intuitive; apart from that, somebody changing c6-v2 DTS will not be aware that he is also indirectly changing another device.

I'd be pleased if we would keep the policy of not doing this at all.

adschm commented:
Unfortunately, this is still incomplete. From a quick look:
- Proper commit message with specifications/flashing instructions, Signed-off-by
- board.d support

https://openwrt.org/submitting-patches

L2jLiga commented:
@adschm i've updated PR

adschm commented:
> According to the GPL source the non-EU variant has different
> GPIOs assigned to some of the LEDs and buttons. The flash
> layout might be different as well.

So, does this include or exclude the RU variant? Please make that clear in the commit message.

adschm commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -760,6 +760,39 @@ static struct device_info boards[] = {
+
+            .first_sysupgrade_partition = "os-image",
+            .last_sysupgrade_partition = "file-system",
+    },
```
Indent is broken here, should be tabs consistently.

L2jLiga replied:
Fixed

L2jLiga committed:
ath79: add support for Archer A6 v2 EU/RU

This patch adds support for TP-Link Archer A6 v2 (EU/RU)

Hardware specification:
- SOC: Qualcomm QCA9563 @ 775MHz
- Flash: GigaDevice GD25Q64CSIG (8MiB)
- RAM: Zentel A3R1GE40JBF (128 MiB DDR2)
- Ethernet: Qualcomm QCA8337N: 4x 1Gbps LAN + 1x 1Gbps WAN
- Wireless:
  - 2.4GHz (bgn) QCA9563 integrated (3x3)
  - 5GHz (ac) Qualcomm QCA9886 (2x2)
- Button: 1x power, 1x reset, 1x wps
- LED: 6x LEDs: power, wlan2g, wlan5g, lan, wan, wps
- UART: There's no UART header on the board

Flash instructions:

Upload
openwrt-ath79-generic-tplink_archer-a6-v2-squashfs-factory.bin
via the router Web interface.

Flash instruction using tftp recovery:

1. Connect the computer to one of the LAN ports of the router
2. Set the computer IP to 192.168.0.66
3. Start a tftp server with the OpenWrt factory image in the
   tftp root directory renamed to ArcherA6v2_tp_recovery.bin.
4. Connect power cable to router, press and hold the reset
   button and turn the router on
5. Keep the reset button pressed until the WPS LED lights up
6. Wait ~150 seconds to complete flashing

According to the GPL source the non-EU variant has different
GPIOs assigned to some of the LEDs and buttons. The flash
layout might be different as well.

Signed-off-by: Andrey Chalkin <L2jLiga@gmail.com>

L2jLiga commented:
> So, does this include or exclude the RU variant? Please make that clear in the commit message.

Let me check

adschm commented:
There appear to be variations in partitions depending on the age of the firmware. This should be investigated more closely.

Kashemir001 commented:
I had a look at six different FWs for RU/EU available at TP-Link website and found three different partition layout schemes among them, here on the picture I tried to visualize them and compare it to the subject PR. Hope that can be useful.
Note that I took partition info from the official FW .bin files starting at address 0x1814, while there are also some partition info at 0x1014, which is different for each FW version file, has less entries compared to 0x1814, and each entry is named "fwup-ptn" instead of just "partition".
![FW_comparsion1](https://user-images.githubusercontent.com/14910998/140059240-d0c79b1e-258a-4233-948c-e64152e74cbb.png)

L2jLiga commented:
Already added in https://github.com/openwrt/firmware-utils/commit/07f78f071075f1b7e866ebea4be821b0ca9e9add
